Hook: get message from .git/COMMIT_EDITMSG instead of console

all3fox opened the pull request:
This pull request is all about making commitplease and husky work together.

Turns out, typicode/husky#30 already attempted to add command line arguments to hooks and it did not work out. So, I decided to modify commitplease instead and made the hook:
  
  1. not care if it is a `./git/commit-msg` or a `node_modules/commitplease/commit-msg-hook.js`
  2. get the commit message from `.git/COMMIT_EDITMSG`

The idea is that the user adds this husky line under "scripts" in the package.json:
```
"commitmsg": "--silent node node_modules/commitplease/commit-msg-hook.js"
```
Then they run `npm install` and actually don't care who creates the hook first. If husky does, then that line from package.json will run commitplease and if commitplease does, then it runs as usual.

The bad thing is this extra configuration bit. Moreover, it is too long. I wanted to make it at least something like:
```
"commitmsg": "--silent node commitplease"
```
but failed :( For it to work, it looks like I need to change  `"main" : "index.js"` to `"main" : "commit-msg-hook.js"` and do something else too.

jzaefferer commented on commit-msg-hook.js:
```diff
@@ -1,4 +1,18 @@
+
+if (!fs.existsSync(fmsg)) {
+  // assume this hook is node_modules/commitplease/commit-msg-hook.js
+  fmsg = path.join(path.resolve(__dirname, '../../.git/' + fname))
```
Could we simplify this using `process.cwd()`? That's what the setup script uses to detect a self-install: https://github.com/jzaefferer/commitplease/blob/master/setup.js#L4

jzaefferer replied:
If that works, we could skip the guessing and access the "absolute" path directly.

jzaefferer replied:
Similar to what that husky issue suggested, `var msg = fs.readFileSync('.git/COMMIT_EDITMSG', 'utf8')`.

jzaefferer commented:
Let's add a "Husky compatibility" (or something like that) section to the readme. To avoid the install order randomness, I'd suggest setting the `nohook: false` option, along with the husky script. That is, when someone is using husky, they should use husky exclusively for dealing with hooks.

As for allowing the main file to also act as hook: Maybe we can merge the logic that accesses the git COMMIT_EDITMSG file into `index.js` and use the separate hook file only when we need to install the hook? `commit-msg-hook.js` would then do nothing except `require('commitplease')()`.

all3fox commented:
> Let's add a "Husky compatibility" (or something like that) section to the readme...

Done, I believe you meant `nohook: true` though. This allows husky to install its hooks.

Changed to `process.cwd()`, `commit-msg-hook.js` is now as short as it used to.

Simplified the `package.json` configuration line to:

```
"scripts": {
    "commitmsg": "node node_modules/commitplease"
}
```

However, the problem is that the silent flag for some reason stopped working for me. When I do:
```
"commitmsg": "--silent node node_modules/commitplease"
```
the husky hook fails. Manually starting up sh and doing the command works:
```
sh
sh-4.3$ npm run --silent commitmsg
```

Without the silent flag the output of npm is confusing when the commit message is malformed. So, I am trying to understand why the flag stopped working. In husky, that looks [like a simple concatenation](https://github.com/typicode/husky/blob/688dc0ea658cc2083e5de07688d4906cbdd805b7/src/index.js#L94) of `npm run ` and whatever is in the `package.json`. Looks like the shell is doing some tricks but I could not yet reach the bottom of it.

jzaefferer commented on index.js:
```diff
@@ -20,3 +21,5 @@ module.exports = function (messageFile) {
   }
 }
+
+module.exports()
```
This seems pointless - let's remove the function and put the code inside on the top level, running it immediately.

jzaefferer commented on commit-msg-hook.js:
```diff
@@ -1,4 +1,4 @@
 
 // `commitplease-original`
-require('commitplease')(process.argv[ 2 ])
+require('commitplease')()
```
I think this should now be just `require('commitplease')`, since that module executes itself.

jzaefferer commented:
Not sure about the `--silent` flag. Maybe someone from husky can help with that?

jzaefferer commented:
Overall this is looking really good. Nice to see that we can get compatibility with husky with so little extra code.

jzaefferer commented on package.json:
```diff
@@ -3,6 +3,7 @@
   "version": "2.3.1",
   "description": "Validates strings as commit messages",
   "main": "index.js",
+  "bin": {"commitplease": "./commit-msg-hook.js"},
```
With this in place, does the readme need another update?

all3fox replied:
Normally, you would comment on some code that needs changing. Once the change gets force-pushed, github marks the comment as "a comment on an outdated diff", so it is clear that the comment has been addressed.

Now, I have changed the readme accordingly but did not change package.json, so force-push does not touch this file. Do I write a comment like "thanks, fixed" here, or do I not respond at all?

jzaefferer replied:
A "fixed" comment works. And I'll avoid commenting on individual commits.

jzaefferer commented:
I've read the discussion in https://github.com/typicode/husky/pull/47 - since the husky people don't seem to know how to fix that either, we could file an issue against the npm cli itself.

all3fox commented:
The "verbosity" of npm seems to be a long-running discussion/problem: npm/npm#5452 and npm/npm#8821 

Various solutions are proposed in npm/npm#5452, I summarized them in my post there and tried to give reasons why they did not work for me.

Many people are complaining about `run-script` being noisy, yet from npm/npm#8821 it is clear that this problem is still only under discussion and might be addressed later this year.

jzaefferer commented:
Alright, looks like there isn't much we can do there. We can point to the npm issue in the readme (when using commitplease with husky) and recommend to use commitplease standalone to avoid that.

all3fox commented:
> We can point to the npm issue in the readme (when using commitplease with husky) and recommend to use commitplease standalone to avoid that.

Done

Now comes the time for #33 and a pull request to jquery. I suggest this:
 1. Merge this PR
 2. Get back to #33, add some unit-tests that show exactly which patterns prohibit those ticket references
 3. Make sure it works, fix code if needed
 4. Merge that PR (unit-tests, maybe code fixes)
 5. Put that to npmjs.com
 6. Test how it installs with latest jquery
 7. Make jquery PR: explain it fixes husky issues and allows to prohibit ticket references

I would say I can do that by Monday, what do you think?

jzaefferer commented on README.md:
````diff
@@ -100,14 +100,16 @@ When using commitplease together with [husky](https://github.com/typicode/husky)
 }
 ```
 
+However, since husky does not use npm in silent mode (and there is no easy way to make it do so: typicode/husky#47 and npm/npm#5452), there will be a lot of additional output when a message fails validation. Therefore, using commitplease alone is recommended.
````
These aren't rendered as links.

jzaefferer commented:
> I would say I can do that by Monday, what do you think?

Sounds great to me! Let's do this!
